client/network/src/protocol/sync/state.rs

```diff
@@ -113,38 +116,77 @@ impl<B: BlockT> StateSync<B> {
 			};
 			log::debug!(target: "sync", "Imported with {} keys", values.len());
 
-			if let Some(last) = values.last().map(|(k, _)| k) {
-				self.last_key = last.clone();
-			}
-
-			for (key, value) in values {
-				self.imported_bytes += key.len() as u64;
-				self.state.push((key, value))
+			let complete = completed == 0;
+			if !complete && !values.update_last_key(completed, &mut self.last_key) {
+				log::debug!(target: "sync", "Error updating key cursor, depth: {}", completed);
 			};
+
+			for values in values.0 {
+				use std::collections::hash_map::Entry;
+				let key_values = if values.parent_storages.len() == 0 {
+					// skip all child key root (will be recalculated on import)
+					values.key_values.into_iter()
+						.filter(|key_value| !well_known_keys::is_child_storage_key(key_value.0.as_slice()))
+						.collect()
+				} else {
+					values.key_values
+				};
+				match self.state.entry(values.parent_storages) {
+					Entry::Occupied(mut entry) => {
+						for (key, value) in key_values {
+							self.imported_bytes += key.len() as u64;
+							entry.get_mut().push((key, value))
+						}
+					},
+					Entry::Vacant(entry) => {
+						for (key, _value) in key_values.iter() {
+							self.imported_bytes += key.len() as u64;
+						}
+						entry.insert(key_values);
+					},
+				}
+			}
 			self.imported_bytes += proof_size;
 			complete
 		} else {
-			log::debug!(
-				target: "sync",
-				"Importing state from {:?} to {:?}",
-				response.entries.last().map(|e| sp_core::hexdisplay::HexDisplay::from(&e.key)),
-				response.entries.first().map(|e| sp_core::hexdisplay::HexDisplay::from(&e.key)),
-			);
-
-			if let Some(e) = response.entries.last() {
-				self.last_key = e.key.clone();
+			let mut complete = true;
+			if self.last_key.len() == 2 && response.entries[0].entries.len() == 0 {
```

Comment:
`response.entries` could be empty here?
Also, why `self.last_key.len() == 2`? Could use some explaining in the comment.

Reply:
It is empty when the whole change set is into a child trie: then we do not have the parent key in parent payload (already define in previous payload). I will add comment.